existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,334 @@
+
+On its own, `ExplicitForAll` doesn't do a lot.
+
+But you frequently need it to enable other extensions. Some of them just won't work without the `forall` quantifier. :)
+
+The most common one is `ScopedTypeVariables`, which allows you to pass a type variable from an outer function to one defined in a where clause.
+
+The code below will not compile because the compiler thinks that the `a` in `example`'s signature and the `a` in `pair`'s are two different types.
+
+```haskell
+example :: a -> [a] -> [a]
+example x rest = pair ++ rest
+  where
+    pair :: [a]
+    pair = [x, x]
+```
+
+To fix it, you need to use the `ScopedTypeVariables` extension and also add a `forall` quantifier to the `example` function, since the extension won't work without it.
+
+```haskell
+example :: forall a. a -> [a] -> [a]
+```
+You might also notice how `forall` was useful with the `TypeApplication` extension.
+Other extensions that somehow benefit from the usage of `ExplicitForAll` are `LiberalTypeSynonyms`, `RankNTypes`, and many more.
+
+## Existential quantification
+
+As I said earlier, besides universally quantified variables, Haskell also supports existential types.
+They are also created with the `forall` keyword.
+
+They can be implemented this way because these two constructions – `(exists x. p x) -> q` and `forall x. (p x -> q)` – are equivalent in terms of first-order predicate logic.
+For a theoretical proof of this statement, you can check this [thread](https://stackoverflow.com/questions/10753073/whats-the-theoretical-basis-for-existential-types?rq=1).
+
+### Usage
+
+In order to proceed further, we'll have to enable the `RankNTypes` extension. This lets us write higher-rank types.
+
+<hr>
+
+**Higher-rank types**
+
+By default, all types are rank-1. In a rank-1 type, a `forall` cannot appear on the left of an arrow (keep in mind that the `->` operator has precedence over `.`)
+
+For example, this function has a rank-1 type:
+
+```haskell
+f :: forall a b. a -> b -> a
+
+-- Or, to make the precedence explicit:
+f :: forall a b. (a -> b -> a)
+```
+
+With the `RankNTypes` extension, we can write rank-2 types – types where `forall` appears to the left of one arrow:
+
+```haskell
+f :: (forall a. Show a => a -> IO ()) -> IO ()
+```
+
+<hr>
+
+Because `forall a` appears to the left of one arrow, `a` is said to be existentially quantified.
+It is also true for every `forall` that is introduced to the left of an odd number of functional arrows.
+But what does this mean?
+
+In contrast to the universal quantifier, the existential one means that the function is itself responsible for instantiating what a type variable should mean. So the result type of existentially quantified variable will become known when we address it at the 'definition-site' (i.e., in the function body).
+This also means that no one can specify its type outside of function.
+
+Let's take a closer look on practical examples.
+
+We will start with functions:
+
+```haskell
```

Comment:
I think you can separate two parts: 

existential quantification in datatypes 
existential quantification in functions